File: graphs (undirected + directed)/BFS.java
package graphs.revision;

public class BFS extends GraphTraversal{
	
	BFS(int vertices) {
		super(vertices);
	}

	public static void main(String[] args) {
		int vertices = 6;
		Graph graph = new Graph(vertices, false);

		graph.insertEdge(1, 2);
		graph.insertEdge(1, 5);
		graph.insertEdge(1, 6);

		graph.insertEdge(2, 5);
		graph.insertEdge(2, 3);

		graph.insertEdge(3, 4);
		graph.insertEdge(4, 5);

		BFS bfs = new BFS(vertices);
		System.out.println("Breadth First Search of the graph");
		bfs.breadthFirstSearch(graph);
	}
	
	@Override
	public void processVertexEarly(int x) {
		System.out.print(x + " ");
	}
	
	@Override
	public void processEdge(int x, int y){
		
	}
	
	@Override
	public void processVertexLate(int y){
		
	}

}

File: algorithms/graphs/BFS.mjs
export default class BreadthFirstSearch {
  constructor(graph) {
    this.graph = graph;
    this.parents = new Array(graph.vertices + 1);
    this.discovered = new Array(graph.vertices + 1);
    this.processed = new Array(graph.vertices + 1);
  }

  search(root) {
    const queue = [];

    queue.push(root);
    this.discovered[root] = true;
    this.parents[root] = -1;

    while (queue.length) {
      let v = queue.shift();

      this.processVertexEarly?.(v);
      this.processed[v] = true;

      let node = this.graph.adjacencyList[v];
      while (node) {
        if (!this.processed[node.y] || this.graph.isDirected) {
          this.processEdge?.(v, node.y, node.weight);
        }

        if (!this.discovered[node.y]) {
          this.discovered[node.y] = true;
          queue.push(node.y);
          this.parents[node.y] = v;
        }
        node = node.next;
      }
      this.processVertextLate?.(v);
    }
  }
}
